### DISCUSSION TITLE
topic `/ir_intensity` contains more than one type

### DISCUSSION THREAD

**Discussion by framlin** [iRobotEducation/create3_docs#267] [Questions] [answered]
Does anybody know, how i can select the type of a topic in rviz2, if a topic hast two types? 
The topic `/ir_intensity` has the types `irobot_create_msgs/msg/IrIntensityVector` and `sensor_msgs/msg/PointCloud` and rviz2 is taking the first one instead of the second for the display PointCloud.
How can I tell the create3 to send `sensor_msgs/msg/PointCloud`-type-messages or rviz2, to display the `irobot_create_msgs/msg/IrIntensityVector`-data?
Or do I miss the point totally, why rviz2 does not display the `/ir_intensity` -data?
Thanks in advance

**Comment by alsora** [comment, marked as answer]
The Create 3 robot publishes data on the `/ir_intensity-data` topic using only `irobot_create_msgs/msg/IrIntensityVector`.
ROS does not support to publish messages of different types on the same topic and, in the more recent releases, it would crash right away if you try to do that.

However, the main point is different.
It sounds like you want to display the `/ir_intensity-data` topic in RViz and unfortunately this is not possible out of the box.
RViz can display only a limited number of topics, i.e. those topics that use message types supported by the RViz plugins.
The  `irobot_create_msgs/msg/IrIntensityVector`, being a "custom" ROS topic developed by iRobot for the Create 3, does not have plugins yet.

You can find here documentation for RViz, which includes instructions on how to write new plugins: https://wiki.ros.org/rviz
It would be a great project if someone would develop a new RViz plugin to display some of the Create 3 topics!

**Reply by framlin** [reply]
Thank you for your answer. So I totally misunderstood, what ros2 and rviz2 are telling me.
Here is my ros2-try to echo the topic:

```
root@verdi:~/create3_ws# ros2 topic echo /ir_intensity
Cannot echo topic '/ir_intensity', as it contains more than one type: [irobot_create_msgs/msg/IrIntensityVector, sensor_msgs/msg/PointCloud]
```

```root@verdi:~/create3_ws# ros2 topic type /ir_intensity
irobot_create_msgs/msg/IrIntensityVector
sensor_msgs/msg/PointCloud
```

From my point of view, that looked as if the topic /ir_intensity hast two types. Why is this not the case and why dose ros2 tell, what it does?

And here is, what rviz2 dumps to the console:

`root@verdi:~/create3_ws# [WARN] [1670430109.385777218] [rviz2]: topic '/ir_intensity' has more than one types associated, rviz will arbitrarily use the type 'irobot_create_msgs/msg/IrIntensityVector' -- all types for the topic: 'irobot_create_msgs/msg/IrIntensityVector' 'sensor_msgs/msg/PointCloud'`


So why do this two applications assume, that the topic /ir_intensity hast two types?

I have one more question:
Some hours ago, calling _'ros2 topic echo /ir_intensity'_, as described in the create3_docs resulted in a message, showing the ir_intensity-data. No it shows, what I have pasted obove? What goes wrong now? I think, I did not change anything between the successfull task earlier this day and the failing one now.

Thanks in advance

**Reply by alsora** [reply]
Can you run this command in a terminal and show me the output?
```
ros2 topic info /ir_intensity --verbose
```

This should show information about the topic.

**Reply by framlin** [reply]
```
root@verdi:~/create3_ws# ros2 topic info /ir_intensity --verbose
Type: irobot_create_msgs/msg/IrIntensityVector

Publisher count: 1

Node name: robot_state
Node namespace: /
Topic type: irobot_create_msgs/msg/IrIntensityVector
Endpoint type: PUBLISHER
GID: 01.10.a6.61.41.f9.b4.95.df.22.5d.41.00.01.b1.03.00.00.00.00.00.00.00.00
QoS profile:
  Reliability: BEST_EFFORT
  Durability: VOLATILE
  Lifespan: 9223372036854775807 nanoseconds
  Deadline: 9223372036854775807 nanoseconds
  Liveliness: AUTOMATIC
  Liveliness lease duration: 9223372036854775807 nanoseconds

Subscription count: 0

```
Indeed, that looks as if there is only one type.

This is after I have restartet my terminals and the galactic-docker-image.
And now, _'ros2 topic echo /ir_intensity'_ shows the ir_intensity-data again ... really strange ....

Thank you for your help!

**Reply by alsora** [reply]
Great, I'm glad it works now.

The explanation I can give is that maybe you had an additional ROS 2 application running on your system and this application was using that other message type you were seeing on that same topic.
An example could also have been trying to publish from ROS 2 command line on the topic.

**Reply by framlin** [reply]
Thank you very much!
So I assume it was rviz2, that confused did that confusing things

**Reply by framlin** [reply]
.... and ..... create 3 is really a great toy ;-) ..........
